File: Reference/DontStarveScript/behaviourtree.lua
require("class")


SUCCESS = "SUCCESS"
FAILED = "FAILED"
READY = "READY"
RUNNING = "RUNNING"


---------------------------------------------------------------------------------------

local function profilewrapvisit(node)
	local oldvisit = node.Visit
	node.Visit = function(self)
		TheSim:ProfilerPush("node: "..tostring(node.name))
		oldvisit(self)
		TheSim:ProfilerPop()
	end
	if node.children then
		for i,child in ipairs(node.children) do
			profilewrapvisit(child)
		end
	end
end

BT = Class(function(self, inst, root)
    self.inst = inst
	--profilewrapvisit(root) -- uncomment for verbose brain profiling!
    self.root = root
end)

function BT:ForceUpdate()
    self.forceupdate = true
end
function BT:Update()

    self.root:Visit()
    self.root:SaveStatus()
    self.root:Step()

    self.forceupdate = false
end

function BT:Reset()
    self.root:Reset()
end

function BT:Stop()
    self.root:Stop()
end

function BT:GetSleepTime()
    if self.forceupdate then
        return 0
    end
    
    return self.root:GetTreeSleepTime()
end

function BT:__tostring()
    return self.root:GetTreeString()
end


---------------------------------------------------------------------------------------

BehaviourNode = Class(function (self, name, children) 
    self.name = name or ""
    self.children = children
    self.status = READY
    self.lastresult = READY
    if children then
        for i,k in pairs(children) do
            k.parent = self
        end
    end
end)

function BehaviourNode:DoToParents(fn)
    if self.parent then
        fn(self.parent)
        return self.parent:DoToParents(fn)
    end
end

function BehaviourNode:GetTreeString(indent)
    indent = indent or ""
    local str = string.format("%s%s>%2.2f\n", indent, self:GetString(), self:GetTreeSleepTime() or 0)
    if self.children then
        for k, v in ipairs(self.children) do
            str = str .. v:GetTreeString(indent .. "   >")
        end
    end
    return str
end

function BehaviourNode:DBString()
    return ""
end

function BehaviourNode:Sleep(t)
    self.nextupdatetime = GetTime() + t 
end

function BehaviourNode:GetSleepTime()
    
    if self.status == RUNNING and not self.children and not self:is_a(ConditionNode) then
        if self.nextupdatetime then
            local time_to = self.nextupdatetime - GetTime()
            if time_to < 0 then
                time_to = 0
            end
            return time_to
        end
        return 0
    end
    
    return nil
end

function BehaviourNode:GetTreeSleepTime()
    
    local sleeptime = nil
    if self.children then
        for k,v in ipairs(self.children) do
            if v.status == RUNNING then
                local t = v:GetTreeSleepTime()
                if t and (not sleeptime or sleeptime > t) then
                    sleeptime = t
                end
            end
        end
    end
    
    local my_t = self:GetSleepTime()
    
    if my_t and (not sleeptime or sleeptime > my_t) then
        sleeptime = my_t
    end
    
    return sleeptime
end

function BehaviourNode:GetString()
    local str = ""
    if self.status == RUNNING then
        str = self:DBString()
    end
    return string.format([[%s - %s <%s> (%s)]], self.name, self.status or "UNKNOWN", self.lastresult or "?", str)
end

function BehaviourNode:Visit()
    self.status = FAILED
end

function BehaviourNode:SaveStatus()
    self.lastresult = self.status
    if self.children then
        for k,v in pairs(self.children) do
            v:SaveStatus()
        end
    end
end

function BehaviourNode:Step()
    if self.status ~= RUNNING then
        self:Reset()
    elseif self.children then
        for k, v in ipairs(self.children) do
            v:Step()
        end
    end
end

function BehaviourNode:Reset()
    if self.status ~= READY then
        self.status = READY
        if self.children then
            for idx, child in ipairs(self.children) do
                child:Reset()
            end
        end
    end
end

function BehaviourNode:Stop()
    if self.OnStop then
        self:OnStop()
    end
    if self.children then
        for idx, child in ipairs(self.children) do
            child:Stop()
        end
    end
end


---------------------------------------------------------------------------------------
DecoratorNode = Class(BehaviourNode, function(self, name, child)
	BehaviourNode._ctor(self, name or "Decorator", {child})
end)

---------------------------------------------------------------------------------------


ConditionNode = Class(BehaviourNode, function(self, fn, name)
    BehaviourNode._ctor(self, name or "Condition")
    self.fn = fn
end)

function ConditionNode:Visit()
    if self.fn() then
        self.status = SUCCESS
    else
        self.status = FAILED
    end
end


---------------------------------------------------------------------------------------


ConditionWaitNode = Class(BehaviourNode, function(self, fn, name)
    BehaviourNode._ctor(self, name or "Wait")
    self.fn = fn
end)

function ConditionWaitNode:Visit()
    if self.fn() then
        self.status = SUCCESS
    else
        self.status = RUNNING
    end
end


---------------------------------------------------------------------------------------


ActionNode = Class(BehaviourNode, function(self, action, name)
    BehaviourNode._ctor(self, name or "ActionNode")
    self.action = action
end)

function ActionNode:Visit()
    self.action()
    self.status = SUCCESS
end


---------------------------------------------------------------------------------------


WaitNode = Class(BehaviourNode, function(self, time)
    BehaviourNode._ctor(self, "Wait")
    self.wait_time = time
end)

function WaitNode:DBString()
    local w = self.wake_time - GetTime()
    return string.format("%2.2f", w)
end

function WaitNode:Visit()
    local current_time = GetTime() 
    
    if self.status ~= RUNNING then
        self.wake_time = current_time + self.wait_time
        self.status = RUNNING
    end
    
    if self.status == RUNNING then
        if current_time >= self.wake_time then
            self.status = SUCCESS
        else
            self:Sleep(current_time - self.wake_time)
        end
    end
    
end


---------------------------------------------------------------------------------------

SequenceNode = Class(BehaviourNode, function(self, children)
    BehaviourNode._ctor(self, "Sequence", children)
    self.idx = 1
end)

function SequenceNode:DBString()
    return tostring(self.idx)
end


function SequenceNode:Reset()
    self._base.Reset(self)
    self.idx = 1
end

function SequenceNode:Visit()
    
    if self.status ~= RUNNING then
        self.idx = 1
    end
    
    local done = false
    while self.idx <= #self.children do
    
        local child = self.children[self.idx]
        child:Visit()
        if child.status == RUNNING or child.status == FAILED then
            self.status = child.status
            return
        end
        
        self.idx = self.idx + 1
    end 
    
    self.status = SUCCESS
end

---------------------------------------------------------------------------------------

SelectorNode = Class(BehaviourNode, function(self, children)
    BehaviourNode._ctor(self, "Selector", children)
    self.idx = 1
end)

function SelectorNode:DBString()
    return tostring(self.idx)
end


function SelectorNode:Reset()
    self._base.Reset(self)
    self.idx = 1
end

function SelectorNode:Visit()
    
    if self.status ~= RUNNING then
        self.idx = 1
    end
    
    local done = false
    while self.idx <= #self.children do
    
        local child = self.children[self.idx]
        child:Visit()
        if child.status == RUNNING or child.status == SUCCESS then
            self.status = child.status
            return
        end
        
        self.idx = self.idx + 1
    end 
    
    self.status = FAILED
end
---------------------------------------------------------------------------------------

NotDecorator = Class(DecoratorNode, function(self, child)
    DecoratorNode._ctor(self, "Not", child)
end)

function NotDecorator:Visit()
	local child = self.children[1]
	child:Visit()
	if child.status == SUCCESS then
		self.status = FAILED
	elseif child.status == FAILED then
		self.status = SUCCESS
	else
		self.status = child.status
	end
end
---------------------------------------------------------------------------------------

FailIfRunningDecorator = Class(DecoratorNode, function(self, child)
    DecoratorNode._ctor(self, "FailIfRunning", child)
end)

function FailIfRunningDecorator:Visit()
	local child = self.children[1]
	child:Visit()
	if child.status == RUNNING then
		self.status = FAILED
	else
		self.status = child.status
	end
end

---------------------------------------------------------------------------------------


LoopNode = Class(BehaviourNode, function(self, children, maxreps)
    BehaviourNode._ctor(self, "Sequence", children)
    self.idx = 1
    self.maxreps = maxreps
    self.rep = 0
end)

function LoopNode:DBString()
    return tostring(self.idx)
end


function LoopNode:Reset()
    self._base.Reset(self)
    self.idx = 1
    self.rep = 0
end

function LoopNode:Visit()
    
    if self.status ~= RUNNING then
        self.idx = 1
        self.rep = 0
    end
    
    local done = false
    while self.idx <= #self.children do
    
        local child = self.children[self.idx]
        child:Visit()
        if child.status == RUNNING or child.status == FAILED then
            if child.status == FAILED then
                --print("EXIT LOOP ON FAIL")
            end
            self.status = child.status
            return
        end
        
        self.idx = self.idx + 1
    end 
    
    self.idx = 1
    
    self.rep = self.rep + 1
    if self.maxreps and self.rep >= self.maxreps then
        --print("DONE LOOP")
        self.status = SUCCESS
    else
        for k,v in ipairs(self.children) do
            v:Reset()
        end
    
    end
end

---------------------------------------------------------------------------------------

RandomNode = Class(BehaviourNode, function(self, children)
    BehaviourNode._ctor(self, "Random", children)
end)


function RandomNode:Reset()
    self._base.Reset(self)
    self.idx = nil
end


function RandomNode:Visit()

    local done = false
    
    if self.status == READY then
        --pick a new child
        self.idx = math.random(#self.children)
        local start = inst.idx
        while true do
        
            local child = self.children[self.idx]
            child:Visit()
            
            if child.status ~= FAILED then
                self.status = child.status
                return
            end
            
            self.idx = self.idx + 1
            if self.idx == #self.children then
                self.idx = 1
            end
            
            if self.idx == start then
                inst.status = FAILED
                return
            end
        end
        
    else
        local child = self.children[self.idx]
        child:Visit()
        self.status = child.status
    end
    
end

---------------------------------------------------------------------------------------    

PriorityNode = Class(BehaviourNode, function(self, children, period)
    BehaviourNode._ctor(self, "Priority", children) 
    self.period = period or 1
end)

function PriorityNode:GetSleepTime()
    if self.status == RUNNING then
        
        if not self.period then
            return 0
        end
        
        
        local time_to = 0
        if self.lasttime then
            time_to = self.lasttime + self.period - GetTime()
            if time_to < 0 then
                time_to = 0
            end
        end
    
        return time_to
    elseif self.status == READY then
        return 0
    end
    
    return nil
    
end


function PriorityNode:DBString()
    local time_till = 0
    if self.period then
       time_till = (self.lasttime or 0) + self.period - GetTime()
    end
    
    return string.format("execute %d, eval in %2.2f", self.idx or -1, time_till)
end


function PriorityNode:Reset()
    self._base.Reset(self)
    self.idx = nil
end

function PriorityNode:Visit()
    
    local time = GetTime()
    local do_eval = not self.lasttime or not self.period or self.lasttime + self.period < time 
    local oldidx = self.idx
    
    
    if do_eval then
        
        local old_event = nil
        if self.idx and self.children[self.idx]:is_a(EventNode) then
            old_event = self.children[self.idx]
        end

        self.lasttime = time
        local found = false
        for idx, child in ipairs(self.children) do
        
            local should_test_anyway = old_event and child:is_a(EventNode) and old_event.priority <= child.priority
            if not found or should_test_anyway then
            
                if child.status == FAILED or child.status == SUCCESS then
                    child:Reset()
                end
                child:Visit()
                local cs = child.status
                if cs == SUCCESS or cs == RUNNING then
                    if should_test_anyway and self.idx ~= idx then
                        self.children[self.idx]:Reset()
                    end
                    self.status = cs
                    found = true
                    self.idx = idx
                end
            else
                
                child:Reset()
            end
        end
        if not found then
            self.status = FAILED
        end
        
    else
        if self.idx then
            local child = self.children[self.idx]
            if child.status == RUNNING then
                child:Visit()
                self.status = child.status
                if self.status ~= RUNNING then
                    self.lasttime = nil
                end
            end
        end
    end
    
end


---------------------------------------------------------------------------------------


ParallelNode = Class(BehaviourNode, function(self, children, name)
    BehaviourNode._ctor(self, name or "Parallel", children)
end)


function ParallelNode:Step()
    if self.status ~= RUNNING then
        self:Reset()
    elseif self.children then
        for k, v in ipairs(self.children) do
            if v.status == SUCCESS and v:is_a(ConditionNode) then
                v:Reset()
            end         
        end
    end
end

function ParallelNode:Visit()
    local done = true
    local any_done = false
    for idx, child in ipairs(self.children) do
        
        if child:is_a(ConditionNode) then
            child:Reset()
        end
        
        if child.status ~= SUCCESS then
            child:Visit()
            if child.status == FAILED then
                self.status = FAILED
                return
            end
        end
        
        if child.status == RUNNING then
            done = false
        else
            any_done = true
        end
        
        
    end

    if done or (self.stoponanycomplete and any_done) then
        self.status = SUCCESS
    else
        self.status = RUNNING
    end    
end

ParallelNodeAny = Class(ParallelNode, function(self, children)
    ParallelNode._ctor(self, children, "Parallel(Any)")
    self.stoponanycomplete = true
end)



---------------------------------------------------------------------------------------


EventNode = Class(BehaviourNode, function(self, inst, event, child, priority)
    BehaviourNode._ctor(self, "Event("..event..")", {child})
    self.inst = inst
    self.event = event
    self.priority = priority or 0

    self.eventfn = function(inst, data) self:OnEvent(data) end
    self.inst:ListenForEvent(self.event, self.eventfn)
    --print(self.inst, "EventNode()", self.event)
end)

function EventNode:OnStop()
    --print(self.inst, "EventNode:OnStop()", self.event)
    if self.eventfn then
        self.inst:RemoveEventCallback(self.event, self.eventfn)
        self.eventfn = nil
    end
end

function EventNode:OnEvent(data)
    --print(self.inst, "EventNode:OnEvent()", self.event)
    
    if self.status == RUNNING then
        self.children[1]:Reset()
    end
    self.triggered = true
    self.data = data
    
    if self.inst.brain then
        self.inst.brain:ForceUpdate()
    end
    
    self:DoToParents(function(node) if node:is_a(PriorityNode) then node.lasttime = nil end end)
    
    --wake the parent!
end

function EventNode:Step()
    self._base.Step(self)
    self.triggered = false
end

function EventNode:Reset()
    self.triggered = false
    self._base.Reset(self)
end

function EventNode:Visit()
    
    if self.status == READY and self.triggered then
        self.status = RUNNING
    end

    if self.status == RUNNING then
        if self.children and #self.children == 1 then
            local child = self.children[1]
            child:Visit()
            self.status = child.status
        else
            self.status = FAILED
        end
    end
    
end

---------------------------------------------------------------

function WhileNode(cond, name, node)
    return ParallelNode
        {
            ConditionNode( cond, name),
            node
        }
end

---------------------------------------------------------------

function IfNode(cond, name, node)
    return SequenceNode
        {
            ConditionNode( cond, name),
            node
        }
end
---------------------------------------------------------------

File: Prototype/behaviourtree.py
# coding=utf-8
# behaviourtree.py
# created on 2020/9/28
# author @zoloypzuo
# usage: behaviourtree
from common.zlogger import logger
from game_timer import now

SUCCESS = "SUCCESS"
FAILED = "FAILED"
READY = "READY"
RUNNING = "RUNNING"


class BT(object):
    """
    Attributes:
    """

    def __init__(self, inst, root):
        self.inst = inst
        self.root = root

    # @logger
    def update(self):
        self.root.visit()
        self.root.save_status()
        self.root.step()

    def reset(self):
        self.root.reset()

    def stop(self):
        self.root.stop()

    def __str__(self):
        # TODO dump tree
        pass


class NodeBase(object):
    """
    递归行为
    """
    import abc

    __metaclass__ = abc.ABCMeta

    @abc.abstractmethod
    def visit(self): pass

    @abc.abstractmethod
    def save_status(self): pass

    @abc.abstractmethod
    def step(self): pass

    @abc.abstractmethod
    def reset(self): pass

    @abc.abstractmethod
    def stop(self): pass


class BehaviourNode(NodeBase):
    """
    Attributes:
        status: /
        last_result: 保存的上次的状态，@save_status

    Event Callbacks:
        on_stop: on_stop()，stop的回调函数
    """

    def __init__(self, name="", children=None):
        self.name = name
        self.children = children or []
        self.status = READY
        self.last_result = READY
        self.parent = None
        if children:
            for child in children:
                child.parent = self

    def visit(self):
        """

        :return:
        """
        self.status = FAILED

    def save_status(self):
        """
        递归地保存状态
        :return:
        """
        self.last_result = self.status
        if self.children:
            for child in self.children:
                child.save_status()

    def step(self):
        """

        :return:
        """
        if self.status != RUNNING:
            self.reset()
        elif self.children:
            for child in self.children:
                child.step()

    def reset(self):
        """
        递归地设置状态为READY
        :return:
        """
        if self.status == READY:
            return
        self.status = READY
        if self.children:
            for child in self.children:
                child.reset()

    def stop(self):
        on_stop = getattr(self, 'on_stop', None)
        on_stop and on_stop()
        if self.children:
            for child in self.children:
                child.stop()

    def do_to_parents(self, fn):
        """
        对父节点链条上的节点应用fn
        :param fn:
        :return:
        """
        if self.parent:
            fn(self.parent)
            self.parent.do_to_parents(fn)

    def __str__(self):
        # TODO dump tree
        pass


class DecoratorNode(BehaviourNode):
    def __init__(self, name, child):
        super(DecoratorNode, self).__init__(name, [child])


class ConditionNode(BehaviourNode):
    def __init__(self, fn, name="Condition"):
        super(ConditionNode, self).__init__(name)
        self.fn = fn

    def visit(self):
        self.status = SUCCESS if self.fn() else FAILED


class ConditionWaitNode(BehaviourNode):
    def __init__(self, fn, name="Wait"):
        super(ConditionWaitNode, self).__init__(name)
        self.fn = fn

    def visit(self):
        self.status = SUCCESS if self.fn() else FAILED


class ActionNode(BehaviourNode):
    def __init__(self, action, name="ActionNode"):
        super(ActionNode, self).__init__(name)
        self.action = action

    def visit(self):
        self.action()
        self.status = SUCCESS


class WaitNode(BehaviourNode):
    """
    第一次访问触发等待开始，到等待时间结束成功
    Attributes:
        wait_time: /
        wake_time: /
    """

    def __init__(self, time):
        super(WaitNode, self).__init__("Wait")
        self.wait_time = time
        self.wake_time = 0

    def visit(self):
        current_time = now()
        if self.status != RUNNING:
            self.wake_time = current_time + self.wait_time
            self.status = RUNNING
        else:
            if current_time >= self.wake_time:
                self.status = SUCCESS
            # else:
            #     ？？？ sleep 《0
            # self.sleep(current_time - self.wake_time)

    def __str__(self):
        return '{:2.2f}'.format(self.wake_time - now())


class SequenceNode(BehaviourNode):
    def __init__(self, children):
        super(SequenceNode, self).__init__("Sequence", children)
        self.index = 0

    def reset(self):
        super(SequenceNode, self).reset()
        self.index = 0

    def visit(self):
        if self.status != RUNNING:
            self.index = 0

        idx = self.index
        len_children = len(self.children)
        while idx < len_children:
            child = self.children[idx]
            child.visit()
            if child.status == RUNNING or child.status == FAILED:
                self.status = child.status
                return
            idx += 1
        self.index = idx
        self.status = SUCCESS


class SelectorNode(BehaviourNode):
    def __init__(self, children):
        super(SelectorNode, self).__init__("Selector", children)
        self.index = 0

    def reset(self):
        super(SelectorNode, self).reset()
        self.index = 0

    def visit(self):
        if self.status != RUNNING:
            self.index = 0
        idx = self.index
        len_children = len(self.children)
        while idx < len_children:
            child = self.children[idx]
            child.visit()
            if child.status == RUNNING or child.status == SUCCESS:
                self.status = child.status
                return
            idx += 1
        self.index = idx
        self.status = FAILED


class NotDecorator(DecoratorNode):
    def __init__(self, child):
        super(NotDecorator, self).__init__("Not", child)

    def visit(self):
        child = self.children[0]
        child.visit()
        if child.status == SUCCESS:
            self.status = FAILED
        elif child.status == FAILED:
            self.status = SUCCESS
        else:
            self.status = child.status


class FailIfRunningDecorator(DecoratorNode):
    def __init__(self, child):
        super(FailIfRunningDecorator, self).__init__("FailIfRunning", child)

    def visit(self):
        child = self.children[0]
        child.visit()
        if child.status == RUNNING:
            self.status = FAILED
        else:
            self.status = child.status


class LoopNode(BehaviourNode):
    def __init__(self, children, maxreps):
        super(LoopNode, self).__init__("Sequence", children)
        self.index = 0
        self.maxreps = maxreps
        self.rep = 0

    def reset(self):
        super(LoopNode, self).reset()
        self.index = 0
        self.rep = 0

    def visit(self):
        if self.status != RUNNING:
            self.index = 0
            self.rep = 0
        idx = self.index
        len_children = len(self.children)
        while idx < len_children:
            child = self.children[idx]
            child.visit()
            if child.status == RUNNING or child.status == FAILED:
                self.status = child.status
                return
            idx += 1
        self.index = 0
        self.rep += 1
        if self.rep >= self.maxreps:
            self.status = SUCCESS
        else:
            for child in self.children:
                child.reset()


class RandomNode(BehaviourNode):
    def __init__(self, children):
        super(RandomNode, self).__init__("Random", children)
        self.index = None

    def reset(self):
        super(RandomNode, self).reset()
        self.index = None

    def visit(self):
        import random
        if self.status == READY:
            self.index = random.randint(0, len(self.children))
            start = self.index
            while True:
                child = self.children[self.index]
                if child.status != FAILED:
                    self.status = child.status

                self.index += 1
                if self.index == len(self.children):
                    self.index = 0
                if self.index == start:
                    self.status = FAILED
                    return
        else:
            child = self.children[self.index]
            child.visit()
            self.status = child.status


class PriorityNode(BehaviourNode):
    def __init__(self, children, period=1.):
        super(PriorityNode, self).__init__("Priority", children)
        self.period = period
        self.index = None
        self.last_time = None

    @logger
    def reset(self):
        super(PriorityNode, self).reset()
        self.index = None

    @property
    def time_till(self):
        """
        离下一次计算的时间
        >0时说明还差n秒进行下一次计算
        <0时visit会执行一次计算
        :return:
        """
        return self.last_time or 0. + self.period - now()

    @logger
    def visit(self):
        """

        :return:
        """
        # 如果没计算过，或者周期到了
        do_eval = not self.last_time or not self.period or self.time_till < 0.
        if do_eval:
            # ---------------------------------------------------
            # 遍历子节点
            # 找到第一个成功的节点
            # ---------------------------------------------------
            old_event = None
            if self.index is not None:
                child = self.children[self.index]
                old_event = child if isinstance(child, EventNode) else None
            self.last_time = now()
            found = False
            for index, child in enumerate(self.children):
                should_test_anyway = old_event and isinstance(child, EventNode) and old_event.priority <= child.priority
                if not found or should_test_anyway:
                    if child.status == FAILED or child.status == SUCCESS:
                        child.reset()
                    child.visit()
                    cs = child.status
                    if cs == SUCCESS or cs == RUNNING:
                        if should_test_anyway and self.index != index:
                            self.children[self.index].reset()
                        self.status = cs
                        found = True
                        self.index = index
                else:
                    child.reset()
            if not found:
                self.status = FAILED
        else:
            # ---------------------------------------------------
            # 运行当前节点
            # ---------------------------------------------------
            if self.index is not None:
                child = self.children[self.index]
                if child.status == RUNNING:
                    child.visit()
                    self.status = child.status
                    if self.status != RUNNING:
                        self.last_time = None


class ParallelNode(BehaviourNode):
    def __init__(self, children, name="Parallel"):
        super(ParallelNode, self).__init__(name, children)
        self.stoponanycomplete = False

    def step(self):
        if self.status != RUNNING:
            self.reset()
        elif self.children:
            for child in self.children:
                if child.status == SUCCESS and isinstance(child, ConditionNode):
                    child.reset()

    def visit(self):
        """

        :return:
        """
        done = True
        any_done = False
        for child in self.children:
            if isinstance(child, ConditionNode):
                child.reset()
            if child.status != SUCCESS:
                child.visit()
                if child.status == FAILED:
                    self.status = FAILED
                    return
            if child.status == RUNNING:
                done = False
            else:
                any_done = True

        self.status = SUCCESS if done or (self.stoponanycomplete and any_done) else RUNNING


class ParallelNodeAny(ParallelNode):
    def __init__(self, children):
        super(ParallelNodeAny, self).__init__(children, "Parallel(Any)")
        self.stoponanycomplete = True


class EventNode(BehaviourNode):
    def __init__(self, inst, event, child, priority=0):
        super(EventNode, self).__init__("Event(\"%s\")" % event, [child])
        self.inst = inst
        self.event = event
        self.priority = priority
        self.eventfn = lambda data: self.on_event(data)
        self.inst.listen_for_event(self.event, self.eventfn)
        self.triggered = False
        self.data = None

    @logger
    def on_event(self, data):
        def parent_handler(node):
            if isinstance(node, PriorityNode):
                node.last_time = None

        if self.status == RUNNING:
            self.children[0].reset()
        self.triggered = True
        self.data = data
        if self.inst.brain:
            self.inst.brain.force_update()

        self.do_to_parents(parent_handler)

    # ---------------------------------------------------
    # override base functions
    # ---------------------------------------------------
    @logger
    def step(self):
        super(EventNode, self).step()
        self.triggered = False

    @logger
    def reset(self):
        self.triggered = False
        super(EventNode, self).reset()

    @logger
    def visit(self):
        if self.status == READY and self.triggered:
            self.status = RUNNING
        if self.status == RUNNING:
            if self.children and len(self.children) == 1:
                child = self.children[0]
                child.visit()
                self.status = child.status
            else:
                self.status = FAILED

    # ---------------------------------------------------
    # event callbacks
    # ---------------------------------------------------
    @logger
    def on_stop(self):
        if self.eventfn:
            self.inst.remove_event_callback(self.event, self.eventfn)
            self.eventfn = None


def WhileNode(cond, name, node):
    return ParallelNode([
        ConditionNode(cond, name),
        node
    ])


def IfNode(cond, name, node):
    return SequenceNode([
        ConditionNode(cond, name),
        node
    ])
